Add cart functionality with GET, POST endpoints

ColinVierheller opened the pull request:
- Implement getCartModel.js to retrieve cart details and items using cartId
- Update getCart.js to handle GET requests for cart details and items, calculate total price
- Implement postCartModel.js to create a new cart for a user if none exists
- Update postCart.js to handle POST requests for creating a new cart using userId from URL parameters
- Implement postCartItemModel.js to add new items to an existing cart, check for existing items
- Update postCartItem.js to handle POST requests for adding items to a cart using cartId, productId, and quantity from URL parameters
- Modify router.js to include routes for GET, POST cart endpoints

ColinVierheller committed:
Add cart functionality with GET, POST endpoints

- Implement getCartModel.js to retrieve cart details and items using cartId
- Update getCart.js to handle GET requests for cart details and items, calculate total price
- Implement postCartModel.js to create a new cart for a user if none exists
- Update postCart.js to handle POST requests for creating a new cart using userId from URL parameters
- Implement postCartItemModel.js to add new items to an existing cart, check for existing items
- Update postCartItem.js to handle POST requests for adding items to a cart using cartId, productId, and quantity from URL parameters
- Modify router.js to include routes for GET, POST cart endpoints

MrChemical1337 reviewed (dismissed):
Looks good

S1onyx commented on test/bruno/cart/getCart.bru:
Please update the bruno file.
there is the wrong route
its actually "http://localhost:3000/api/cart/id"
an not with parameters

S1onyx commented on backend/apis/cart/getCart.js:
```diff
@@ -0,0 +1,33 @@
+    const cartData = await getCartModel(cartId);
+    if (!cartData || cartData.items.length === 0) {
+      res.writeHead(404);
+      return res.end(JSON.stringify({ error: 'Cart items not found' }));
```
maybe its better when you don't return an error if the cart is existing but empty?!

S1onyx replied:
please show the cartDetails, 
response the item section empty
totalPrice = 0

S1onyx commented on test/bruno/cart/postCart.bru:
postCart and postCartItems looks the same to me in bruno
please make sure that the tests are correct. 
Save them in bruno and push it to the repo

S1onyx commented on backend/models/cart/postCartModel.js:
```diff
@@ -0,0 +1,29 @@
+
+async function postCartModel(userId) {
+  try {
+    // Überprüfen, ob bereits ein Warenkorb für den Benutzer existiert
```
good!

S1onyx commented on backend/apis/cart/postCart.js:
```diff
@@ -0,0 +1,20 @@
+      res.writeHead(409, { 'Content-Type': 'application/json' });
+      return res.end(JSON.stringify({ error: 'Cart already exists for this user', cartId: result.cartId }));
+    }
+
```
maybe add an error if the client wants to create an cart for an user that doesn't exists

S1onyx commented on backend/apis/cart/postCartItem.js:
```diff
@@ -0,0 +1,20 @@
+    if (result.itemExists) {
+      res.writeHead(409, { 'Content-Type': 'application/json' });
+      return res.end(JSON.stringify({ error: 'Item already exists in the cart', itemId: result.itemId }));
+    }
```
maybe add an error if the product id doesn't exists and don't throw the server error?

S1onyx commented on backend/apis/cart/postCartItem.js:
Please think about how to handle the case if the item in the cart already exists. 
Maybe it is an option to update the cart and build the sum of the quantity
Or you throw an explicit error - than please Crete a new subtask for an API endpoint that we can update (PUT) an existing cartItem with a new quantity

S1onyx commented on backend/models/cart/postCartItemModel.js:
```diff
@@ -0,0 +1,29 @@
+
+    // Erstellen eines neuen Warenkorb-Artikels
+    const newItemQuery = await db.query(
+      `INSERT INTO cart_items (cart_id, product_id, quantity) VALUES ($1, $2, $3) RETURNING id`,
```
nice solution that you returning the id directly, vor dev its great to have this!!!
maybe in a production product this is not the way to go but for our purpose its great!

ColinVierheller committed:
- Bruno Files fixed
- getCart works now for empty and non existing carts

ColinVierheller committed:
Added some fixes
unknown users and carts now throw errors
items are added up if they exist in the cart already

ColinVierheller committed:
test

ColinVierheller committed:
Fixed negative quantity problem

ColinVierheller committed:
Added deleteCartItem
-> deletes item fully from cart

ColinVierheller committed:
Added deleteCart functionality, deletes the entire
cart, if cart doesn't exist, throw error

ColinVierheller committed:
Added putQuantity to update quantity after adding
product to cart

ColinVierheller committed:
Merge branch 'dev' into feature-cart-api-endpoints

MrChemical1337 reviewed (dismissed):
After reviewing the files (20 😢) thoroughly I can approve the Request 👍

Commit:
Merge remote-tracking branch 'origin/dev' into feature-cart-api-endpoints

# Conflicts:
#	backend/routes/router.js

Commit:
commit conflict

S1onyx commented:
Please update all endpoint, so that only the cart owner can delete, post and put cart_items.
please dont parse the user_id manually, its important, that the endpoint reads the user_id from the jwt in the cookie
it's not allowed to create an cart or cart_items for an other user.
Pease update the endpoint and the router for ALL endpoints

ColinVierheller committed:
Merge remote-tracking branch 'origin/dev' into feature-cart-api-endpoints

ColinVierheller committed:
Edited deleteCart, deleteCartItem, and putQuantity
- User can only edit his own cart (Routing edited)
- Cart is automatically created when product is added to empty cart
- No unnecessary parameters (json body implemented)

ColinVierheller committed:
Merge branch 'dev' into feature-cart-api-endpoints

S1onyx reviewed (approved):
lgtm
